libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,584 @@
+        // end by resetting the state of the cleanup methods.
+        reset_simplification();
+        reset_pruning();
+        return true;
+    }
+    return false;
+}
+
+// restart simplification algorithm, should be called whenever a new point is added
+void AP_SafeRTL::restart_simplification()
+{
+    _simplify_complete = false;
+    _simplify_stack_last_index = -1;
+    _simplify_bitmask.setall();
+}
+
+// reset simplification algorithm so that it will re-check all points in the path
+// should be called if the existing path is altered for example when a loop as been removed
+void AP_SafeRTL::reset_simplification()
+{
+    _simplify_clean_until = 0;
+    restart_simplification();
+}
+
+// restart pruning algorithm, should be called whenever a new point is added
+void AP_SafeRTL::restart_pruning()
+{
+    _prune_complete = false;
+    _prune_current_i = _prune_clean_until;
+    _prune_min_j = _prune_clean_until+2;
+    _prunable_loops_last_index = -1; // clear the loops that we've recorded
+}
+
+// reset pruning algorithm so that it will re-check all points in the path
+// should be called if the existing path is altered for example when a loop as been removed
+void AP_SafeRTL::reset_pruning()
+{
+    _prune_clean_until = 0;
+    restart_pruning();
+}
+
+void AP_SafeRTL::zero_points_by_simplify_bitmask()
+{
+    for (int16_t i = 0; i <= _last_index; i++) {
+        if (!_simplify_bitmask.get(i)) {
+            _simplify_clean_until = MIN(_simplify_clean_until, i-1);
+            if (_path[i] != Vector3f(0.0f, 0.0f, 0.0f)) {
+                log_action(SRTL_POINT_SIMPLIFY, _path[i]);
+                _path[i].zero();
+            }
+        }
+    }
+}
+
+/**
+*   Only prunes loops until $points_to_delete points have been removed. It does not necessarily prune all loops.
+*/
+void AP_SafeRTL::zero_points_by_loops(int16_t points_to_delete)
+{
+    int16_t removed_points = 0;
+    for (int16_t i = 0; i <= _prunable_loops_last_index; i++) {
+        prune_loop_t l = _prunable_loops[i];
+        _prune_clean_until = MIN(_prune_clean_until, l.start_index-1);
+        for (int16_t j = l.start_index; j < l.end_index; j++) {
+            // zero this point if it wasn't already zeroed
+            if (_path[j] != Vector3f(0.0f, 0.0f, 0.0f)) {
+                log_action(SRTL_POINT_PRUNE, _path[j]);
+                _path[j].zero();
+            }
+        }
+        _path[(int16_t)((l.start_index+l.end_index)/2.0)] = l.halfway_point;
+        removed_points += l.end_index - l.start_index - 1;
+        if (removed_points > points_to_delete) {
+            return;
+        }
+    }
+}
+
+/**
+*  Removes all NULL points from the path, and shifts remaining items to correct position.
+*  The first item will not be removed.
+*/
+void AP_SafeRTL::remove_empty_points()
+{
+    int16_t i = 0;
+    int16_t j = 0;
+    int16_t removed = 0;
+    while (++i <= _last_index) { // never removes the first item. This should always be {0,0,0}
```

Comment:
It's elsewhere implied that the first item isn't always {0, 0, 0} ArduCopter/control_safe_rtl.cpp:19 is it always going to be {0, 0, 0}? If so get_point() should be moved to a private helper, and control_safe_rtl should just compare against the length of the point it's looking at.